File: Backtracking/Baekjoon/14888.js
const fs = require("fs");
const file = "./input.txt";
// const file = '/dev/stdin';
const input = fs.readFileSync(file).toString().trim().split("\n");

const n = parseInt(input[0]);
const numbers = input[1].split(" ").map(Number);
let [plus, minus, multiple, divide] = input[2].split(" ").map(Number);

let max = -1000000001;
let min = 1000000001;
solve(1, numbers[0]);
function solve(i, answer) {
  if (i === n) {
    max = Math.max(answer, max);
    min = Math.min(answer, min);
    return;
  }
  if (plus > 0) {
    plus -= 1;
    solve(i + 1, answer + numbers[i]);
    plus += 1;
  }
  if (minus > 0) {
    minus -= 1;
    solve(i + 1, answer - numbers[i]);
    minus += 1;
  }
  if (multiple > 0) {
    multiple -= 1;
    solve(i + 1, answer * numbers[i]);
    multiple += 1;
  }
  if (divide > 0) {
    divide -= 1;
    solve(i + 1, parseInt(answer / numbers[i]));
    divide += 1;
  }
}

// ✅ js에서는 +0 -0이 출력됨 ㅜ.ㅜ
console.log(max ? max : 0);
console.log(min ? min : 0);

File: Backtracking/Baekjoon/14888.py
n = int(input())
numbers = list(map(int, input().split(' ')))
plus, minus, multiple, divide = map(int, input().split(' '))

max_answer = -1000000001
min_answer = 1000000001

def solve(i, answer):
    global plus
    global minus
    global multiple
    global divide
    global max_answer
    global min_answer
    if(i==n):
        max_answer = max(answer, max_answer)
        min_answer = min(answer, min_answer)
        return
    if(plus>0):
        plus-=1
        solve(i+1, answer + numbers[i])
        plus+=1
    if(minus>0):
        minus-=1
        solve(i+1, answer - numbers[i])
        minus+=1
    if(multiple>0):
        multiple-=1
        solve(i+1, answer * numbers[i])
        multiple+=1
    if(divide>0):
        divide-=1
        if(answer<0): solve(i+1, -(-1*answer//numbers[i]))
        else: solve(i+1, answer//numbers[i])
        divide+=1
    
    
solve(1, numbers[0])
print(max_answer)
print(min_answer)
